src/box/lua/tuple.c

```diff
@@ -705,14 +728,172 @@ static const struct luaL_Reg lbox_tuple_meta[] = {
+	mpstream_flush(&stream);
+	size_t size = region_used(region) - region_svp;
+	const char *data = xregion_join(region, size);
+	luamp_decode(L, luaL_msgpack_default, &data);
+	region_truncate(region, region_svp);
+}
+
+/*
+ * Next function of `{i}pairs`.
+ */
+static int
+luaT_tuple_format_iterator_next(struct lua_State *L)
+{
+	struct tuple_format *fmt = luaT_check_tuple_format(L, 1);
+	uint32_t field_cnt = tuple_format_field_count(fmt);
+	if (field_cnt == 0) {
+		lua_pushnil(L);
+		return 1;
+	}
+
+	if (lua_isnil(L, 2)) {
+		lua_pushnumber(L, 1);
+		push_tuple_field_format(L, fmt, 0);
+		return 2;
+	}
+
+	uint32_t i = luaL_touint64(L, 2);
+	if (i >= field_cnt) {
+		lua_pushnil(L);
+		return 1;
+	}
+	lua_pushnumber(L, i + 1);
+	push_tuple_field_format(L, fmt, i);
+	return 2;
+}
+
+/*
+ * Returns (i, {name, type}) pairs.
+ */
+static int
+lbox_tuple_format_pairs(struct lua_State *L)
+{
+	luaT_check_tuple_format(L, 1);
+	lua_rawgeti(L, LUA_REGISTRYINDEX, luaT_tuple_format_iterator_next_ref);
+	lua_insert(L, 1);
+	lua_pushnil(L);
+	return 3;
+}
+
+/*
+ * Returns an array tuple field formats.
+ */
+static int
+lbox_tuple_format_serialize(struct lua_State *L)
+{
+	struct tuple_format *fmt = luaT_check_tuple_format(L, 1);
+	const char *fmt_mp = fmt->mp;
+	luamp_decode(L, luaL_msgpack_default, &fmt_mp);
+	luaL_findtable(L, LUA_GLOBALSINDEX, "box.internal.space", 1);
+	lua_getfield(L, -1, "denormalize_format");
+	lua_remove(L, -2);
+	lua_pushvalue(L, -2);
+	lua_call(L, 1, 1);
+	return 1;
+}
+
+int
+lbox_push_tuple_format(struct lua_State *L, struct tuple_format *format)
+{
+	struct tuple_format **ptr = (struct tuple_format **)
+		luaL_pushcdata(L, CTID_STRUCT_TUPLE_FORMAT_PTR);
+	*ptr = format;
+	tuple_format_ref(format);
+	return 1;
+}
+
+static int
+lbox_tuple_format_new(struct lua_State *L)
+{
+	assert(CTID_STRUCT_TUPLE_FORMAT_PTR != 0);
+	int top = lua_gettop(L);
+	if (top == 0)
+		return lbox_push_tuple_format(L, tuple_format_runtime);
+	assert(top == 1 && lua_istable(L, 1));
+	uint32_t count = lua_objlen(L, 1);
+	if (count == 0)
+		return lbox_push_tuple_format(L, tuple_format_runtime);
+	struct ibuf *buf = cord_ibuf_take();
+	size_t unused;
+	const char *fmt = luaT_tuple_encode_on_lua_ibuf(L, 1, &unused, buf);
+	if (fmt == NULL) {
+		cord_ibuf_drop(buf);
+		luaT_error(L);
+	}
+	struct region *region = &fiber()->gc;
+	size_t region_svp = region_used(region);
+	struct field_def *fields;
+	uint32_t field_count;
+	int rc = field_def_array_decode(&fmt, &fields, &field_count, region);
+	cord_ibuf_drop(buf);
+	if (rc != 0) {
+		region_truncate(region, region_svp);
+		return luaT_error(L);
+	}
+	struct tuple_format *format =
+		runtime_tuple_format_new(fields, field_count, false);
+	region_truncate(region, region_svp);
+	if (format == NULL)
+		return luaT_error(L);
+	return lbox_push_tuple_format(L, format);
+}
+
+static const struct luaL_Reg lbox_tuple_format_meta[] = {
+	{"__gc", lbox_tuple_format_gc},
```

Comment:
How does it work?

```
tarantool> for i = 1, 1e5 do box.tuple.format.new({{name = 'field' .. i, type = 'decimal'}}) end
---
- error: 'Tuple format limit reached: 65536'
...

tarantool> collectgarbage()
---
- 0
...

tarantool> collectgarbage()
---
- 0
...

tarantool> box.tuple.new({1}, {format = {{name = 'v', type = 'integer'}}})
---
- error: 'Tuple format limit reached: 65536'
...

tarantool> collectgarbage()
---
- 0
...

tarantool> box.tuple.new({1}, {format = {{name = 'v', type = 'integer'}}})
---
- error: 'Tuple format limit reached: 65536'
...
```

I expect that it will be possible to create new format after GC.

Reply:
This was caused because Lua tuple format objects didn't get garbage collected: after adding these two lines it works

```diff
diff --git a/src/box/lua/tuple.c b/src/box/lua/tuple.c
index e075f65de..56cf99ced 100644
--- a/src/box/lua/tuple.c
+++ b/src/box/lua/tuple.c
@@ -843,6 +843,8 @@ lbox_push_tuple_format(struct lua_State *L, struct tuple_format *format)
                luaL_pushcdata(L, CTID_STRUCT_TUPLE_FORMAT_PTR);
        *ptr = format;
        tuple_format_ref(format);
+       lua_pushcfunction(L, lbox_tuple_format_gc);
+       luaL_setcdatagc(L, -2);
        return 1;
 }
```

But according to LuaJIT documentation [^1]:
> The __gc metamethod only applies to struct/union types and performs an implicit [ffi.gc()](https://luajit.org/ext_ffi_api.html#ffi_gc) call during creation of an instance.

So I don't understand why an implicit `ffi.gc` is needed here.

[^1]: https://luajit.org/ext_ffi_api.html

Reply:
Is it possible to cover this case with a test?

Reply:
Yeah, added your loop example.